main/util.py

```diff
@@ -214,16 +216,38 @@ def create_name_from_email(email):
   return re.sub(r'_+|-+|\.+|\++', ' ', email.split('@')[0]).title()
 
 
-def password_hash(user_db, password):
+def password_hash(password, salt):
   m = hashlib.sha256()
-  m.update(user_db.key.urlsafe())
-  m.update(user_db.created.isoformat())
-  m.update(m.hexdigest())
   m.update(password.encode('utf-8'))
-  m.update(config.CONFIG_DB.salt)
+  m.update(salt)
   return m.hexdigest()
 
 
+def password_salt(n_bytes=32):
+  return binascii.hexlify(os.urandom(n_bytes))
+
+
+# Taken from Django Source Code
+# Helps prevent timing attacks
+# See http://carlos.bueno.org/2011/10/timing.html
+def constant_time_compare(val1, val2):
```

Comment:
yes.. you are right :)

but at the moment you can have only 8 attempts (https://github.com/gae-init/gae-init/blob/master/main/config.py#L32) before the reCAPTCHA will kick in.. so I don't see any reason to optimise that part and I think it will make more sense with native apps rather than webapps as precise timing depending on many other things especially when milliseconds count :)

Also regarding the custom login in general there is a very nice post by Jeff Atwood on how they did on Discourse https://blog.codinghorror.com/the-god-login/ and one thing that I'm considering adding in the near future is the server side time delay after consecutive wrong attempts.. to get rid of the reCAPTCHA :)